File: app/src/main/java/com/unidospelovolei/domain/scheduling/TeamDraft.kt
package com.unidospelovolei.domain.scheduling

import com.unidospelovolei.domain.model.Genero
import com.unidospelovolei.domain.model.Player
import com.unidospelovolei.domain.model.Team
import com.unidospelovolei.domain.model.TeamRoster
import kotlin.math.abs
import kotlin.random.Random

data class ElencoPassado(
    val jogadores: List<String>,
    val peso: Double,
)

class HistoricoDeDuplas private constructor(
    private val pesos: Map<String, Map<String, Double>>,
) {
    fun peso(
        a: String,
        b: String,
    ): Double = pesos[a]?.get(b) ?: 0.0

    fun penalidade(
        jogadorId: String,
        elenco: List<Player>,
    ): Double = elenco.sumOf { peso(jogadorId, it.id) }

    fun repeticoes(elencos: List<List<Player>>): Double =
        elencos.sumOf { elenco ->
            var soma = 0.0
            elenco.forEachIndexed { i, a ->
                for (j in i + 1 until elenco.size) soma += peso(a.id, elenco[j].id)
            }
            soma
        }

    companion object {
        val VAZIO = HistoricoDeDuplas(emptyMap())

        fun de(elencos: List<ElencoPassado>): HistoricoDeDuplas {
            val acumulado = mutableMapOf<String, MutableMap<String, Double>>()
            elencos.forEach { elenco ->
                elenco.jogadores.forEachIndexed { i, a ->
                    for (j in i + 1 until elenco.jogadores.size) {
                        val b = elenco.jogadores[j]
                        acumulado.getOrPut(a) { mutableMapOf() }.merge(b, elenco.peso, Double::plus)
                        acumulado.getOrPut(b) { mutableMapOf() }.merge(a, elenco.peso, Double::plus)
                    }
                }
            }
            return HistoricoDeDuplas(acumulado)
        }
    }
}

object TeamDraft {
    const val ALVO_POR_GENERO: Int = 2

    const val JOGADORES_POR_TIME: Int = ALVO_POR_GENERO * 2

    const val TENTATIVAS_PADRAO: Int = 240

    private const val PESO_HISTORICO_NA_ESCOLHA = 2.5
    private const val RUIDO_NA_ESCOLHA = 1.5

    private const val PESO_AMPLITUDE = 1.0
    private const val PESO_REPETICAO = 2.5
    private const val PESO_GENERO = 12.0

    fun distribute(
        players: List<Player>,
        teams: List<Team>,
        historico: HistoricoDeDuplas = HistoricoDeDuplas.VAZIO,
        random: Random = Random.Default,
        tentativas: Int = TENTATIVAS_PADRAO,
    ): List<TeamRoster> {
        val timesEmOrdem = teams.sortedWith(compareBy({ it.ordem }, { it.nome.lowercase() }))
        if (timesEmOrdem.isEmpty()) return emptyList()

        val ativos = players.filter { it.ativo }
        if (ativos.isEmpty()) return timesEmOrdem.map { TeamRoster(it, emptyList()) }

        var melhor: List<List<Player>> = montar(ativos, timesEmOrdem.size, historico, random)
        var melhorCusto = custo(melhor, historico)
        repeat((tentativas - 1).coerceAtLeast(0)) {
            val candidato = montar(ativos, timesEmOrdem.size, historico, random)
            val custo = custo(candidato, historico)
            if (custo < melhorCusto) {
                melhorCusto = custo
                melhor = candidato
            }
        }

        return timesEmOrdem.mapIndexed { indice, time ->
            TeamRoster(
                team = time,
                players =
                    melhor[indice].sortedWith(
                        compareByDescending<Player> { it.skillLevel }.thenBy { it.nome.lowercase() },
                    ),
            )
        }
    }

    private fun montar(
        ativos: List<Player>,
        totalTimes: Int,
        historico: HistoricoDeDuplas,
        random: Random,
    ): List<List<Player>> {
        val capacidade = capacidades(ativos.size, totalTimes, random)
        val porGenero = Genero.entries.associateWith { genero -> ativos.filter { it.genero == genero } }
        val vagas = cotasPorGenero(capacidade, porGenero.mapValues { it.value.size }, random)

        val elencos = List(totalTimes) { mutableListOf<Player>() }

        listOf(Genero.FEMININO, Genero.MASCULINO).forEach { genero ->
            val restante = vagas.getValue(genero)

            val fila = porGenero.getValue(genero).shuffled(random).sortedByDescending { it.skillLevel }
            fila.forEach { jogador ->
                val destino = melhorDestino(jogador, elencos, restante, capacidade, historico, random)
                elencos[destino] += jogador
                restante[destino] -= 1
            }
        }

        return elencos.map { it.toList() }
    }

    private fun capacidades(
        total: Int,
        totalTimes: Int,
        random: Random,
    ): IntArray {
        val capacidade = IntArray(totalTimes) { total / totalTimes }
        (0 until totalTimes).shuffled(random).take(total % totalTimes).forEach { capacidade[it] += 1 }
        return capacidade
    }

    private fun cotasPorGenero(
        capacidade: IntArray,
        totais: Map<Genero, Int>,
        random: Random,
    ): Map<Genero, IntArray> {
        val n = capacidade.size
        val feminino = IntArray(n) { minOf(ALVO_POR_GENERO, capacidade[it]) }
        val masculino = IntArray(n) { capacidade[it] - feminino[it] }
        val mulheres = totais[Genero.FEMININO] ?: 0

        var soma = feminino.sum()
        while (soma > mulheres) {
            val time = escolher(random, (0 until n).filter { feminino[it] > 0 }) { feminino[it] } ?: break
            feminino[time] -= 1
            masculino[time] += 1
            soma -= 1
        }
        while (soma < mulheres) {
            val time = escolher(random, (0 until n).filter { masculino[it] > 0 }) { -feminino[it] } ?: break
            feminino[time] += 1
            masculino[time] -= 1
            soma += 1
        }

        return mapOf(Genero.FEMININO to feminino, Genero.MASCULINO to masculino)
    }

    private fun escolher(
        random: Random,
        candidatos: List<Int>,
        chave: (Int) -> Int,
    ): Int? {
        if (candidatos.isEmpty()) return null
        val melhor = candidatos.maxOf(chave)
        return candidatos.filter { chave(it) == melhor }.random(random)
    }

    private fun melhorDestino(
        jogador: Player,
        elencos: List<List<Player>>,
        vagasDoGenero: IntArray,
        capacidade: IntArray,
        historico: HistoricoDeDuplas,
        random: Random,
    ): Int {
        val disponiveis =
            vagasDoGenero.indices.filter { vagasDoGenero[it] > 0 }.ifEmpty {
                elencos.indices.filter { elencos[it].size < capacidade[it] }.ifEmpty {
                    val menor = elencos.minOf { it.size }
                    elencos.indices.filter { elencos[it].size == menor }
                }
            }

        return disponiveis.minBy { time ->
            elencos[time].sumOf { it.skillLevel } +
                PESO_HISTORICO_NA_ESCOLHA * historico.penalidade(jogador.id, elencos[time]) +
                random.nextDouble() * RUIDO_NA_ESCOLHA
        }
    }

    private fun custo(
        elencos: List<List<Player>>,
        historico: HistoricoDeDuplas,
    ): Double {
        val forcas = elencos.map { elenco -> elenco.sumOf { it.skillLevel } }
        val amplitude = (forcas.max() - forcas.min()).toDouble()
        val desequilibrio =
            elencos.sumOf { elenco ->
                val mulheres = elenco.count { it.genero == Genero.FEMININO }
                abs(mulheres - minOf(ALVO_POR_GENERO, elenco.size))
            }.toDouble()

        return PESO_AMPLITUDE * amplitude +
            PESO_REPETICAO * historico.repeticoes(elencos) +
            PESO_GENERO * desequilibrio
    }
}

File: web/src/domain/teamDraft.ts
import type { Player, Team, TeamRoster } from './models';
import { defaultRandom, minBy, randomElement, shuffled, sortedBy, type Rng } from './random';

export interface ElencoPassado {
  jogadores: string[];
  peso: number;
}

export class HistoricoDeDuplas {
  static readonly VAZIO = new HistoricoDeDuplas(new Map());

  private constructor(private readonly pesos: Map<string, Map<string, number>>) {}

  static de(elencos: readonly ElencoPassado[]): HistoricoDeDuplas {
    const acumulado = new Map<string, Map<string, number>>();
    const somar = (a: string, b: string, peso: number): void => {
      const linha = acumulado.get(a) ?? new Map<string, number>();
      linha.set(b, (linha.get(b) ?? 0) + peso);
      acumulado.set(a, linha);
    };

    elencos.forEach((elenco) => {
      for (let i = 0; i < elenco.jogadores.length; i++) {
        for (let j = i + 1; j < elenco.jogadores.length; j++) {
          somar(elenco.jogadores[i], elenco.jogadores[j], elenco.peso);
          somar(elenco.jogadores[j], elenco.jogadores[i], elenco.peso);
        }
      }
    });
    return new HistoricoDeDuplas(acumulado);
  }

  peso(a: string, b: string): number {
    return this.pesos.get(a)?.get(b) ?? 0;
  }

  penalidade(jogadorId: string, elenco: readonly Player[]): number {
    return elenco.reduce((soma, jogador) => soma + this.peso(jogadorId, jogador.id), 0);
  }

  repeticoes(elencos: readonly (readonly Player[])[]): number {
    return elencos.reduce((total, elenco) => {
      let soma = 0;
      for (let i = 0; i < elenco.length; i++) {
        for (let j = i + 1; j < elenco.length; j++) soma += this.peso(elenco[i].id, elenco[j].id);
      }
      return total + soma;
    }, 0);
  }
}

export const ALVO_POR_GENERO = 2;
export const JOGADORES_POR_TIME = ALVO_POR_GENERO * 2;
export const TENTATIVAS_PADRAO = 240;

const PESO_HISTORICO_NA_ESCOLHA = 2.5;
const RUIDO_NA_ESCOLHA = 1.5;

const PESO_AMPLITUDE = 1.0;
const PESO_REPETICAO = 2.5;
const PESO_GENERO = 12.0;

const forcaDe = (elenco: readonly Player[]): number =>
  elenco.reduce((soma, jogador) => soma + jogador.skillLevel, 0);

function capacidades(total: number, totalTimes: number, random: Rng): number[] {
  const capacidade = new Array<number>(totalTimes).fill(Math.floor(total / totalTimes));
  const indices = shuffled(
    Array.from({ length: totalTimes }, (_, i) => i),
    random,
  ).slice(0, total % totalTimes);
  indices.forEach((indice) => {
    capacidade[indice] += 1;
  });
  return capacidade;
}

function escolher(random: Rng, candidatos: readonly number[], chave: (i: number) => number): number | null {
  if (candidatos.length === 0) return null;
  const melhor = Math.max(...candidatos.map(chave));
  return randomElement(
    candidatos.filter((candidato) => chave(candidato) === melhor),
    random,
  );
}

function cotasPorGenero(
  capacidade: readonly number[],
  totalDeMulheres: number,
  random: Rng,
): { feminino: number[]; masculino: number[] } {
  const n = capacidade.length;
  const feminino = capacidade.map((vagas) => Math.min(ALVO_POR_GENERO, vagas));
  const masculino = capacidade.map((vagas, i) => vagas - feminino[i]);
  const indices = Array.from({ length: n }, (_, i) => i);

  let soma = feminino.reduce((total, valor) => total + valor, 0);
  while (soma > totalDeMulheres) {
    const time = escolher(
      random,
      indices.filter((i) => feminino[i] > 0),
      (i) => feminino[i],
    );
    if (time === null) break;
    feminino[time] -= 1;
    masculino[time] += 1;
    soma -= 1;
  }
  while (soma < totalDeMulheres) {
    const time = escolher(
      random,
      indices.filter((i) => masculino[i] > 0),
      (i) => -feminino[i],
    );
    if (time === null) break;
    feminino[time] += 1;
    masculino[time] -= 1;
    soma += 1;
  }

  return { feminino, masculino };
}

function melhorDestino(
  jogador: Player,
  elencos: readonly Player[][],
  vagasDoGenero: readonly number[],
  capacidade: readonly number[],
  historico: HistoricoDeDuplas,
  random: Rng,
): number {
  const indices = Array.from({ length: elencos.length }, (_, i) => i);

  let disponiveis = vagasDoGenero
    .map((_, i) => i)
    .filter((i) => vagasDoGenero[i] > 0);
  if (disponiveis.length === 0) {
    disponiveis = indices.filter((i) => elencos[i].length < capacidade[i]);
  }
  if (disponiveis.length === 0) {
    const menor = Math.min(...elencos.map((elenco) => elenco.length));
    disponiveis = indices.filter((i) => elencos[i].length === menor);
  }

  return minBy(
    disponiveis,
    (time) =>
      forcaDe(elencos[time]) +
      PESO_HISTORICO_NA_ESCOLHA * historico.penalidade(jogador.id, elencos[time]) +
      random.nextDouble() * RUIDO_NA_ESCOLHA,
  );
}

function montar(
  ativos: readonly Player[],
  totalTimes: number,
  historico: HistoricoDeDuplas,
  random: Rng,
): Player[][] {
  const capacidade = capacidades(ativos.length, totalTimes, random);
  const masculinos = ativos.filter((jogador) => jogador.genero === 'masculino');
  const femininas = ativos.filter((jogador) => jogador.genero === 'feminino');
  const vagas = cotasPorGenero(capacidade, femininas.length, random);

  const elencos: Player[][] = Array.from({ length: totalTimes }, () => []);

  const distribuir = (jogadores: readonly Player[], restante: number[]): void => {
    const fila = sortedBy(shuffled(jogadores, random), (jogador) => -jogador.skillLevel);
    fila.forEach((jogador) => {
      const destino = melhorDestino(jogador, elencos, restante, capacidade, historico, random);
      elencos[destino].push(jogador);
      restante[destino] -= 1;
    });
  };

  distribuir(femininas, vagas.feminino);
  distribuir(masculinos, vagas.masculino);

  return elencos;
}

function custo(elencos: readonly Player[][], historico: HistoricoDeDuplas): number {
  const forcas = elencos.map(forcaDe);
  const amplitude = Math.max(...forcas) - Math.min(...forcas);
  const desequilibrio = elencos.reduce((soma, elenco) => {
    const mulheres = elenco.filter((jogador) => jogador.genero === 'feminino').length;
    return soma + Math.abs(mulheres - Math.min(ALVO_POR_GENERO, elenco.length));
  }, 0);

  return (
    PESO_AMPLITUDE * amplitude +
    PESO_REPETICAO * historico.repeticoes(elencos) +
    PESO_GENERO * desequilibrio
  );
}

export function distribute(
  players: readonly Player[],
  teams: readonly Team[],
  historico: HistoricoDeDuplas = HistoricoDeDuplas.VAZIO,
  random: Rng = defaultRandom(),
  tentativas: number = TENTATIVAS_PADRAO,
): TeamRoster[] {
  const timesEmOrdem = sortedBy(
    teams,
    (time) => time.ordem,
    (time) => time.nome.toLowerCase(),
  );
  if (timesEmOrdem.length === 0) return [];

  const ativos = players.filter((jogador) => jogador.ativo);
  if (ativos.length === 0) return timesEmOrdem.map((team) => ({ team, players: [] }));

  let melhor = montar(ativos, timesEmOrdem.length, historico, random);
  let melhorCusto = custo(melhor, historico);
  for (let i = 0; i < Math.max(tentativas - 1, 0); i++) {
    const candidato = montar(ativos, timesEmOrdem.length, historico, random);
    const custoDoCandidato = custo(candidato, historico);
    if (custoDoCandidato < melhorCusto) {
      melhorCusto = custoDoCandidato;
      melhor = candidato;
    }
  }

  return timesEmOrdem.map((team, indice) => ({
    team,
    players: sortedBy(
      melhor[indice],
      (jogador) => -jogador.skillLevel,
      (jogador) => jogador.nome.toLowerCase(),
    ),
  }));
}
